### ISSUE TITLE
Expand Plugin to Any Supported Meta Type in SDR with Certain Exceptions

### ISSUE THREAD

**Issue by mcarvin8** [mcarvin8/sfdx-decomposer-plugin#39]
https://github.com/mcarvin8/sfdx-decomposer-plugin/tree/feat/metadata-expansion

Allow any meta suffix to be used as input if they exist in the [SDR Metadata Registry](https://github.com/forcedotcom/source-deploy-retrieve/blob/main/src/registry/metadataRegistry.json). Parent meta suffixes are only supported using the `registryAccess.getTypeBySuffix` method so children meta types (like custom fields) will not be supported.

Block specific metadata types with certain [adapter strategies](https://github.com/forcedotcom/source-deploy-retrieve/blob/main/HANDBOOK.md#adapters) like Apex classes, triggers, etc. Block Custom Fields completely since this is already decomposed by SDR.

These strategies so far have been added to the branch code to block:
```
      if (
        metadataTypeEntry.strategies?.adapter &&
        ['matchingContentFile', 'digitalExperience', 'mixedContent', 'bundle'].includes(
          metadataTypeEntry.strategies.adapter
        )
      ) {
        this.error(
          `Metadata types with ${metadataTypeEntry.strategies.adapter} strategies are not eligible for this plugin: ${metadataTypeToRetrieve}.`
        );
      }
```

SDR has certain unique ID elements which could be used to name decomposed files, but it does not have all the ones this plugin has which I deem necessary for the 20 types already supported. If no unique ID elements are provided/found, it will default to using the SHA-256 hash of the decomposed element contents to name the decomposed file (update added in a previous release). This could be the default assumption for naming decomposed files unless we work with Salesforce to thoroughly update SDR with unique ID elements. The `--purge` flag I added to `decompose` would be useful to run in most cases so you remove any pre-existing decomposed files that may conflict with a newer decomposed file with a different SHA hash.

@AllanOricil / @scolladon - I wouldn't mind getting your thoughts on:
1. How do we sufficiently test this for all supported meta types in SDR? This update on `feat/metadata-expansion` so far hasn't broken functionality for the previously 20 supported meta types in `force-app` per the E2E test. But testing this for all supported meta types in SDR will certainly be tough unless Salesforce has any open-source repositories with sample meta files for all metadata types? I'd rather not release this until I can accurately say this works for all meta types in SDR.
2. Pros and Cons on expanding this plugin from the 20 hard-coded types to all types in SDR except for ones we choose to block (like custom objects)

**Comment by AllanOricil** [comment]
@scolladon is the guy you are looking for to help here 👉

**Comment by mcarvin8** [comment]
One positive - Found this bug while testing the above lol https://github.com/mcarvin8/sfdx-decomposer-plugin/commit/3259d855e89d495d34d4847ac21ba6f54f62db82

Pushed this for v2.1.5

**Comment by AllanOricil** [comment]
Ok, I git curious and I remember I used this guy once https://github.com/certinia/apex-mdapi/tree/master/apex-mdapi%2Fsrc%2Fdocuments%2FMetadataServicePatcher

You can probably reverter engineer those models to schemas you can then use.

**Comment by AllanOricil** [comment]
If 3 can't be done, and you want to use 2 instead of 1, I suggest making its output a standalone package. On every API release, trigger an automation to generate/update new schemas. Then version your packages using the same api version you are scrapping so that there is a 1:1 match always.

**Comment by mcarvin8** [comment]
Thanks @AllanOricil , I'll give these a shot!

**Comment by mcarvin8** [comment]
Yea I'm curious how the Salesforce documentation gets generated with example meta files. I have found some mistakes with XML syntax in the docs, so there may need to be some work done to correct those if Salesforce has a tool 🤣

**Comment by AllanOricil** [comment]
Weird. Do you have a case where you reported it and they confirmed? @mcarvin8

**Comment by mcarvin8** [comment]
Nah I didn't make a case report but I can do so on Monday. 

This page has an issue with an extra `>` I'm pretty sure - https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_marketingappextension.htm

And I do recall one of the very first metadata types I implemented add a weird XML bug too (workflow, permission set, profile, or flow). I need to look at those docs again because it was weird one like `<fullname>something<name>` where the start tag and end tag had different text.

**Comment by AllanOricil** [comment]
If these schemas are public someone can finally create a DML for Salesforce to enable data modeling using code instead of UI or metadata!!!

**Comment by mcarvin8** [comment]
It's workflow - https://developer.salesforce.com/docs/atlas.en-us.api_meta.meta/api_meta/meta_workflow.htm

`  <fullName>Field_Update</name>`

**Comment by AllanOricil** [comment]
@mcarvin8 you can still scrape schemas it using the tables. I think they are more reliable than the examples. The examples seem to be written by hand, while the tables are following the same pattern. Tables are all under `data colSort`, and metadata types are a sibilling tag, right above. Child data is always linked and you can extract the anchor from it.

**Comment by scolladon** [comment]
Hello there, I try to catch up and bring my 2 cents here

> Ask @scolladon (your insider haha) if he can ask someone at Salesforce what is the source of truth they are using to generate that Metadata API doc. Maybe there is a java package where they store all those types? Maybe this data is currently private, but I don't see a reason for it not becoming public, since its doc is public.

IMHO it is an automated process, metadata developer create their new metadata type via the metadata API. They do not write anything inside sfdx. I think it is an automated process that parse the metadata API wsdl and then create the json for sfdx. Then sfdx engineer map this to SDR specificities if needed.

------------

> I wasn't sure at the time I found this where they tracked documentation issues so I just fixed it and moved on but I can raise a case on Monday through my work account.

The documentation is not written by developer but by CCX Engineer (Customer Content Experience Engineer) because we treat the documentation as a showcase and we want to provide the best cohesive experience across every documentation and releases (that's a very hard task).
The documentation is not taken from source code (no litteral programming). This is why sometimes there are misprint I guess.

---------

> How do we sufficiently test this for all supported meta types in SDR

This is a very simple question with a lot of very hard potential solutions. Being completely exhaustive would be a tremendous effort to generate the first version and also to maintain it across releases.
What we did in `sgd` we tested every "kind" at least one, this way we know the way we treat the kind is correct and if there is an issue it is either a bug for the kind or a bad kind detection.
But we have our MetadataRegistry, because historically we were dealing with more metadata type than the SDR (I don't know if it is still true, need to spike).
So, what I propose here is to test the kind detection with unit test and at least one element kind treatment in integration / e2e test. When there is an issue with a kind you can then fix the issue and add a test for this specific kind at the right level (unit / integration / e2e)

-----------

> You could also publish these schemas here [SchemaStore/schemastore](https://github.com/SchemaStore/schemastore)

I think it is a great idea, but it should be owned by `sfdx` IMHO

-------------

> Or we can avoid starting metadata deploy jobs that have mistakes in the metada!!! I think even sfdx does not do this check yet. Right? 

I think it is possible to start a deployment with wrongly formatted xml file. The deployment will fail with weird issues but it will start IMHO. The schema and the metadata consistency are validated inside Salesforce server when trying to apply the deployment content.
Your point is we could validate schema before deploying and so detect syntax issue earlier (we could even build a husky pre-commit hook to avoid committing wrongly formatted files) and I really like it @AllanOricil 👍 

-------------------

Unique ID could be the "key" like `fullName` value for most of the xml element in file (there are exception with Profiles though).
The file could be contained inside a sub type folder (like for CustomObject children per exemple).
Here is the pattern: `folder/metadata-type/metadata-name/child-metadata-type/child-metadata-identifier.child-metadata-type-meta.xml`

exemple: force-app/main/default/profiles/Admin.profile-meta.xml
=>
- force-app/main/default/profiles/Admin/classAccesses/TestClass.classAccesses-meta.xml
- etc

**Comment by mcarvin8** [comment]
Thanks for the update @scolladon .

In terms of testing "kinds" with this plugin, I think the "kinds' inside these XMLs are:
1. Nested elements inside the XML (1st level in the XML, I'm thinking of ways to allow decomposing down the 1st level, but that will take some time. This plugin is stable with decomposing 1 level in the original XML)
2. Leaf elements inside the XML
3. Array of Nested elements inside the XML (i.e. multiple nested elements have the same tag names `<name>`)
4. Array of Leaf elements inside the XML (i.e. multiple leaf elements have the same tag names `<name>text</name>...<name>other text</name> `)

With yesterday's release (2.1.5), I fixed case/kind #4 which I caught testing Custom Applications (a random type I picked after expanding to SDR on the new feature branch) but case #4 doesn't seem to be possible with the 20 meta types supported by the latest version of this plugin on `main`.

Since 2.1.5 is testing SAT on the 4 "kinds" above and the `feat/metadata-expansion` branch is testing SAT on "newer" types like custom apps, I think the "kinds" are tested sufficiently by the current test if I add some custom app files to the repo.

In terms of the Unique ID elements used to name decomposed files, the default ID element on 2.1.5 is `fullName`, so that would be set for all metadata types. If `fullName` isn't found and we don't list other potential unique ID elements, then it defaults to using the SHA-256 hash of the element contents to name the decomposed file. So I think that's a safe bet to leave as-is if we expand to the majority of meta types in SDR.

Besides ensuring the testing is sufficient, I still want to take a further look into SDR to determine what other metadata types I should block from this plugin if expanding the scope. 

This code I added to the feature branch handles the cases that stood out to me (metadata types like apex classes which have a meta XML file) which I would want to block from this plugin, but perhaps there's more.
```
      if (
        metadataTypeEntry.strategies?.adapter &&
        ['matchingContentFile', 'digitalExperience', 'mixedContent', 'bundle'].includes(
          metadataTypeEntry.strategies.adapter
        )
      ) {
        this.error(
          `Metadata types with ${metadataTypeEntry.strategies.adapter} strategies are not eligible for this plugin: ${metadataTypeToRetrieve}.`
        );
      }
```

So perhaps I thought of the testing wrong. Even though it would be created to test all meta types in SDR that I do not add blockers for, the effort to either scrape and clean all the XML data for each type (since the documentation does have multiple XMLs for each type in some instances) would be a big one.

Maybe the most efficient path is to thoroughly test the 4 "kinds" of XML elements to decompose and recompose I listed above. Then once this is ready for release, I just update the README to obviously add disclaimers stating that this plugin is expanded to the majority of SDR types , but to use caution and report any bugs to this project for specific types as users find them.

🤷  Not sure if that makes sense or not. There isn't too much in terms of XML elements to decompose and recompose (leafs and nested elements) and I'm pretty confident the fix pushed for v2.1.5 will handle the majority of cases. But there could be weird cases I'm not accounting for. I've added support for CDATA characters and fixed how comments were processed in translation files, but there could be other weird characters that get added to these XMLs I'm not accounting for in newer meta types.
